Add explanation for Component and ComponentContents (#4)
neoforged/Documentation: docs/misc/components.md

* `Component#empty()` - creates empty component
* `Component#keybind(String name)` - creates component which is represented as name of a Keyboard key for passed keybind.
* `Component#nbt(String path, boolean interpreting, Optional<Component> separator, DataSource dataSource)` - creates component for representing nbt data specified by `path` inside `dataSource`
* `Component#score(String name, String objective)` - creates component for representing `objective`'s score of entity specified by [entity selector][selectors] `name`.
* `Component#selector(String pattern, Optional<Component> separator)` - creates component for displaying names of entities selected by [entity selector][selectors] `pattern`.

Applying Style
==============

`Style` specifies how component should look like, such as bold, italic, color, custom font, etc.
Style is immutable objects and any modification to already existing style will instead create new instance. Its constructor is `private` so you'll have to copy and reconfigure `Style.EMPTY`.

Style's default configurations are `null` instead of concrete value. This means not explicitly specified configuration may be overridden.

> After configuring style to your preference, you can apply it to your component with either `MutableComponent#setStyle` for overwriting previous style, or `#withStyle` for only applying non-null configurations to already existing style.

Reviewer: The `#withStyle` needs a better description because the statement is technically incorrect depending on which version of the method you are using.

Here's an example of how you can stylize your component:
```java
MutableComponent text = Component.literal("Hello!"); // Creates MutableComponent wrapping literal "Hello!"

Style blueItalic = Style.EMPTY // Copies empty style and sets color to blue and makes it italic
        .withColor(0x0000FF)
        .withItalic(true);
Style red = Style.EMPTY // Copies empty style and sets color to red
        .withColor(0xFF0000);
Style bold = Style.EMPTY // Copies empty style and makes it bold
        .withBold(true);

Style doubleLines = Style.EMPTY // Copies empty style and makes it both underlined and strikethrough
        .withUnderlined(true)

>         .withStrikethrough(true);

Reviewer: Put comments on top of code since otherwise people will have to scroll to read it.

text.setStyle(blue); // Sets style of the text to be blue and italic

> text.withStyle(red).withStyle(bold).withStyle(doubleLines); // Overwrites blue and italic style to be red, bold, underlined, and strikethrough

Reviewer: This needs an image. It is not obviously clear what is happening.

```
This results in red, bold text with two lines.

Chaining Components
===================

`MutableComponent` can have additional components chained as siblings with `MutableComponent#append`. Chained components can be retrieved with `MutableComponent#getSiblings`.

`Component` stores its siblings like a tree, and its style is inherited to its siblings.
Below example will progressively apply style on top of each other:
```java
Style bold = Style.EMPTY.withBold(true);
Style italic = Style.EMPTY.withItalic(true);
Style strike = Style.EMPTY.withStrikethrough(true);
Style underlined = Style.EMPTY.withUnderlined(true);

MutableComponent a = Component.literal("a");
MutableComponent b = Component.literal("b").setStyle(bold);
MutableComponent c = Component.literal("c").setStyle(italic);
MutableComponent d = Component.literal("d").setStyle(strike);
MutableComponent e = Component.literal("e").setStyle(underlined);
a.append(b);
b.append(c);
c.append(d);

> d.append(e);

Reviewer: Comments are necessary.

```

> [Here's how it looks like in-game][style_annotated]

Reviewer: I would use a better image. You can get one using a command block and title text or draw a black ground and render the component on top.

`ComponentContents`
===================

Component's text contents are handled by `ComponentContents`, it holds data and defines how to represent it as text.
Here's the list of predefined `ComponentContents`:
* `LiteralContents(String text)` - represents passed string literal as is.
* `TranslatableContents(String translationKey, Object[] args)` - represents passed `translationkey` as localized text. `args` is inserted in place of format specifiers such as `%s` and `%1$s`, read [formatting] for more details.
* `KeybindContents(String name)` - represents keybind name such as `tutorial.move.title` as name of Keyboard Key.
* `NbtContents(String path, boolean interpreting, Optional<Component> separator, DataSource dataSource)` - represents data from `dataSource` specified by `path`. `separator` will be used when there are multiple entities.
* `ScoreContents(String name, String objective)` - represents `objectives`'s score of entities selected by [entity selector][selectors] `name`  .

> * `SelectorContents(String pattern, Optional<Component> separator)` - represents entity selector specified by [entity selector][selectors] `pattern` as a list of selected entities' display names. `separator` will be used when there are multiple entities.

Reviewer: You already mentioned the static helpers above, this section seems redundant and probably should be mentioned above instead.

You may have discovered that parameters of constructors are same as static helper methods of `Component`.
In fact, those helper methods simply create contents with passed parameter and returns `MutableComponent` wrapping it.

Text Formatting

> ==========

Reviewer: Not sure whether this needs to be a section since only translatable contents are supported here.

Text formatting is the process of inserting data as text into predefined larger text, such as displaying coordinate with x, y, z annotation, showing number with respectful units alongside, etc.
Usually special notation called **format specifiers** are used for indicating where a text can be inserted into.

In Minecraft, only `TranslatableContents` supports formatting, and it uses two types of format specifiers: `%s` and `%1$s`, `%2$s`, `%3$s`. `TranslatableContents`'s `args` are inserted in place of format specifiers.
`%s` is replaced with elements of `args` in order they appear, i.e., First `%s` is replaced with a first element of `args`, and so on.
`%n$s` is positional specifier, they can specify which element to replace them with number `n`.
Here's an example on how format specifier works in practice:
* Formatting `%s %s %s` with `[1, 2, 3]` results in `1 2 3`
* Formatting `%1$s, %2$s, %3$s` with `[1, 2, 3]` results in `1, 2, 3`

> * Formatting `%3$s - %2$s - %1$s` with `[1, 2, 3]` results in `3 - 2 - 1`

Reviewer: Needs an example to explain its usecase.

Text Components
==================

A `Component` is a holder for text which can be formatted and chained with other components via its subtype `MutableComponent`. A component can be created using one of the available static helpers:

| Method Name    | Description                                                                                                             |
|----------------|-------------------------------------------------------------------------------------------------------------------------|

> | `literal`      | it creates a component which simply wraps passed in text.                                                               |
> | `nullToEmpty`  | it's same as `#literal` except it creates empty component if null has been passed                                       |
> | `translatable` | it creates a component which is represented as localized text to user, read [internationalization] for more details.    |
> | `empty`        | it creates an empty component                                                                                           |
> | `keybind`      | it creates a component which is represented as name of current Keyboard key of passed keybind.                          |
> | `nbt`          | it creates a component for representing nbt data specified by `path` inside `dataSource`                                |
> | `score`        | it creates a component for representing `objective`'s score of entity specified by [entity selector][selectors] `name`. |
> | `selector`     | it creates a component for displaying list of names of entities selected by [entity selector][selectors] `pattern`.     |

Reviewer: ```suggestion
| `literal`      | it creates a component which simply wraps the passed in text.                                                               |
| `nullToEmpty`  | it's the same as `#literal` except it creates an empty component if null has been passed                                       |
| `translatable` | it creates a component which is represented as localized text to user, read [internationalization] for more details.    |
| `empty`        | it creates an empty component                                                                                           |
| `keybind`      | it creates a component which is represented as the name of current keyboard key of the passed key mapping.                          |
| `nbt`          | it creates a component for representing nbt data specified by `path` inside `dataSource`                                |
| `score`        | it creates a component for representing the `objective`'s score of an entity specified by the [entity selector][selectors] `name`. |
| `selector`     | it creates a component for displaying the list of names of entities selected by the [entity selector][selectors] `pattern`.     |
```

Add a link to the key mapping page.

> A component's text contents are represented by `ComponentContents`. Notably, subtype `TranslatableContents` not only supports [localization][internalization] but also [text formatting][formatting].

Reviewer: ```suggestion
A component's text contents are represented by `ComponentContents`. Notably, the subtype `TranslatableContents` not only supports [localization][internalization] but also [text formatting][formatting].
```

// Create components with style
MutableComponent red = Component.litearl("red ").withStyle(Style.EMPTY.withColor(0xFF0000));
MutableComponent blue = Component.literal("blue ").withStyle(Style.EMPTY.withColor(0x0000FF));
MutableComponent bold = Component.literal("bold ").withStyle(Style.EMPTY.withBold(true));

// Structure created components in the same way as the image
red.append(first).append(blue).append(seventh);
blue.append(second).append(third).append(bold);
bold.append(fourth).append(fifth).append(sixth);
```
![style_annotated]

Text Formatting
---------------

Text formatting is the process of inserting data as text into predefined larger text. It can be used for displaying coordinates, showing unit measurements, etc. **Format specifiers** are used for indicating where a text can be inserted.

`TranslatableContents` allows two types of format specifiers: `%s` and `%n$s`. The component uses the second parameter onwards, denoted as `args` , for holding what object to insert in place of a format specifier.

`%s` is replaced with elements of `args` in order they appear, i.e., the first `%s` is replaced with the first element of `args`, and so on.

> `%n$s` is positional specifier; they can specify which element in `args` will replace the specifier via the number `n`.

Reviewer: ```suggestion
`%n$s` is positional specifier; each positional specifier can denote which element in `args` will replace the specifier via the number `n`.
```

`TranslatableContents` allows two types of format specifiers: `%s` and `%n$s`. The component uses the second parameter onwards, denoted as `args` , for holding what object to insert in place of a format specifier.

`%s` is replaced with elements of `args` in order they appear, i.e., the first `%s` is replaced with the first element of `args`, and so on.
`%n$s` is positional specifier; each positional specifier can denote which element in `args` will replace the specifier via the number `n`.
* Formatting `x:%s y:%s z:%s` with `[1, 2, 3]` as `args` results in `x:1 y:2 z:3`
* Formatting `Time: %1$s ms` with `17` as `args` results in `Time: 17 ms`
* Formatting `Player name: %2$s, HP: %1$s` with `[10.2, Dev]` as `args` results in `Player name: Dev, HP: 10.2`

Any `Component` element within `args` will be transformed into a formatted text string.

[internalization]: ../concepts/internationalization.md
[selectors]: https://minecraft.fandom.com/wiki/Target_selectors
[red_hello]: /img/component_red_hello.png
[style_annotated]: /img/component_style_annotated.png
[formatting]: #text-formatting
[tree]: /img/component_graph.png

> [keymapping]: ../misc/keymappings.md

Reviewer: ```suggestion
[keymapping]: ./keymappings.md
```

The page is already in the `misc` directory.

| `nullToEmpty`  | it's same as `#literal` except it creates empty component if null has been passed                                       |
| `translatable` | it creates a component which is represented as localized text to user, read [internationalization] for more details.    |
| `empty`        | it creates an empty component                                                                                           |
| `keybind`      | it creates a component which is represented as name of current Keyboard key of passed keybind.                          |
| `nbt`          | it creates a component for representing nbt data specified by `path` inside `dataSource`                                |
| `score`        | it creates a component for representing `objective`'s score of entity specified by [entity selector][selectors] `name`. |
| `selector`     | it creates a component for displaying list of names of entities selected by [entity selector][selectors] `pattern`.     |

Component's text contents are represented by `ComponentContents`, which holds data and defines how to represent it as text.

> These helpers create `ComponentContents`, and wrap it in a `MutableComponent`.

Reviewer: Redundant sentence, can be removed.

Author: I don't know which one you're referring to, is it `These helpers create ComponentContents, and wrap it in a MutableComponent.`?

Reviewer: Yes.

Notably, subtype `TranslatableContents` not only supports [localization][internalization] but also [text formatting][formatting].

Applying Style
--------------

Components can be formatted (e.g., bold, click actions, color) via `Style`s.

> `Style`s are immutable, creating new `Style` each time when modified.

Reviewer: ```suggestion
`Style`s are immutable, creating a new `Style` each time when modified.
```

> You can reconfigure `Style.EMPTY` to your preferences.

Reviewer: ```suggestion
The empty style `Style#EMPTY` can be used as a base for configuration.
```

> Multiple styles can be merged together with `#applyTo(Style other)`, `other` will override all empty configurations of `this`

Reviewer: ```suggestion
Multiple styles can be merged together with `#applyTo(Style other)`; `other` will override all non-configured of the current object.
```

> After configuring a style, it can be applied to a component with either `MutableComponent#setStyle` for overwriting, or `#withStyle` for merging with previous one:

Reviewer: ```suggestion
After configuring a style, it can be applied to a component with either `MutableComponent#setStyle` for overwriting, or `#withStyle` for merging:
```

```java
// Creates MutableComponent wrapping literal "Hello!"
MutableComponent text = Component.literal("Hello!");

// Copies empty style and sets color to blue and makes it italic
Style blueItalic = Style.EMPTY
        .withColor(0x0000FF)
        .withItalic(true);

// Copies empty style and sets color to red
Style red = Style.EMPTY 
        .withColor(0xFF0000);

// Copies empty style and makes it bold
Style bold = Style.EMPTY
        .withBold(true);

// Copies empty style and makes it both underlined and strikethrough
Style doubleLines = Style.EMPTY
        .withUnderlined(true)
        .withStrikethrough(true);

// Sets style of the text to be blue and italic
text.setStyle(blueItalic);

// Overwrites blue and italic style to be red, bold, underlined, and strikethrough
text.withStyle(red).withStyle(bold).withStyle(doubleLines);
```

> This results in red, bold text with two lines as shown in the image:

Reviewer: ```suggestion
This creates a red, bold text with two lines:
```

![red_hello]

Chaining Components
-------------------

> `MutableComponent` can have additional components chained as siblings with `MutableComponent#append`. Chained components can be retrieved with `MutableComponent#getSiblings`.

Reviewer: ```suggestion
`MutableComponent#append` can chain multiple components together. Chained components can be retrieved with `MutableComponent#getSiblings`.
```

> `Component` stores its siblings like a tree and traversed in preorder; the parent style is merged with those of its siblings.

Reviewer: ```suggestion
`Component` stores its siblings like a tree and is traversed in preorder; the parent style is merged with those of its siblings.
```

![tree]

The code below will create a component with the same structure in the above example:
```java
// Create text only components
MutableComponent first = Component.literal("first ");
MutableComponent second = Component.literal("second ");
MutableComponent third = Component.literal("third ");
MutableComponent fourth = Component.literal("fourth ");
MutableComponent fifth = Component.literal("fifth ");
MutableComponent sixth = Component.literal("sixth ");
MutableComponent seventh = Component.literal("seventh ");

// Create components with style
MutableComponent red = Component.litearl("red ").withStyle(Style.EMPTY.withColor(0xFF0000));
MutableComponent blue = Component.literal("blue ").withStyle(Style.EMPTY.withColor(0x0000FF));
MutableComponent bold = Component.literal("bold ").withStyle(Style.EMPTY.withBold(true));

// Structure created components in the same way as the image
red.append(first).append(blue).append(seventh);
blue.append(second).append(third).append(bold);
bold.append(fourth).append(fifth).append(sixth);
```

> Here's how it looks like in-game

Reviewer: Unnecessary since it's immediately follows.

![style_annotated]

Text Formatting
---------------

Text formatting is the process of inserting data as text into predefined larger text.

> It can be used for displaying coordinate with x, y, z annotation, showing number with respectful units alongside, etc.

Reviewer: ```suggestion
It can be used for displaying coordinates, showing unit measurements, etc.
```

> Usually special notation called **format specifiers** are used for indicating where a text can be inserted into.

Reviewer: ```suggestion
**Format specifiers** are used for indicating where a text can be inserted.
```

> `TranslatableContents` supports text formatting and uses two types of format specifiers: `%s` and `%n$s`.

Reviewer: ```suggestion
`TranslatableContents` allows two types of format specifiers: `%s` and `%n$s`.
```

> It has object array `args` for holding what to insert in place of format specifiers.

Reviewer: ```suggestion
The component uses the second parameter onwards, denoted as `args` , for holding what object to insert in place of a format specifier.
```

> `%s` is replaced with elements of `args` in order they appear, i.e., First `%s` is replaced with a first element of `args`, and so on.

Reviewer: ```suggestion
`%s` is replaced with elements of `args` in order they appear, i.e., the first `%s` is replaced with the first element of `args`, and so on.
```

`%n$s` is positional specifier; they can specify which element will replace them with number `n`.
* Formatting `x:%s y:%s z:%s` with `[1, 2, 3]` as `args` results in `x:1 y:2 z:3`
* Formatting `Time: %1$s ms` with `17` as `args` results in `Time: 17 ms`
* Formatting `Player name: %2$s, HP: %1$s` with `[10.2, Dev]` as `args` results in `Player name: Dev, HP: 10.2`

> `args`'s elements may be `Component`s that will be inserted into the resulting formatted text with all their attributes preserved.

Reviewer: ```suggestion
When an element within `args` is a `Component`, the component will be formatted as normal.
```

Author: I don't think this sufficiently conveys how the component will be formatted, what does it mean by "normal"?

Reviewer: I was thinking as normal or as specified since we already defined how a component is formatted above.

Reviewer: That's a fair point on the "normal" sentence. Hmm, it still feels a bit clunky to read though.

What about this?
```
Any `Component` element within `args` will be transformed into a formatted text string.
```

| `literal`      | it creates component which simply wraps passed in text.                                                               |
| `nullToEmpty`  | it's same as `#literal` except it creates empty component if null has been passed                                     |
| `translatable` | it creates component which is represented as localized text to user, read [internationalization] for more details.    |
| `empty`        | it creates empty component                                                                                            |
| `keybind`      | it creates component which is represented as name of current Keyboard key of passed keybind.                          |
| `nbt`          | it creates component for representing nbt data specified by `path` inside `dataSource`                                |
| `score`        | it creates component for representing `objective`'s score of entity specified by [entity selector][selectors] `name`. |
| `selector`     | it creates component for displaying list of names of entities selected by [entity selector][selectors] `pattern`.     |

> These helpers create [`ComponentContents`][content] and wrap it in `MutableComponent`

Author: Would "These helpers create `ComponentContents`, and wrap it in a `MutableComponent`." be good?

> These helpers create [`ComponentContents`][content] and wrap it in `MutableComponent`

Reviewer: Looks fine to me.

Applying Style
--------------

Components can be formatted (e.g., bold, click actions, color) via `Style`s.
They're immutable objects which create copy of themselves instead when modified.
You can reconfigure `Style.EMPTY` to your preferences.

`Style`'s configurations can have empty values and multiple styles can be merged together with `#applyTo(Style other)`,
with `other` overriding all empty configurations of `this`

After configuring style to your preference,
you can apply it to your component with either `MutableComponent#setStyle` for overwriting,
or `#withStyle` for merging with previous one.

> Here's an example of how you can stylize your component:

Reviewer: Not sure if this sentence is needed since it's followed up below.

```java
// Creates MutableComponent wrapping literal "Hello!"
MutableComponent text = Component.literal("Hello!");

// Copies empty style and sets color to blue and makes it italic
Style blueItalic = Style.EMPTY
        .withColor(0x0000FF)
        .withItalic(true);

// Copies empty style and sets color to red
Style red = Style.EMPTY 
        .withColor(0xFF0000);

// Copies empty style and makes it bold
Style bold = Style.EMPTY
        .withBold(true);

// Copies empty style and makes it both underlined and strikethrough
Style doubleLines = Style.EMPTY
        .withUnderlined(true)
        .withStrikethrough(true);

// Sets style of the text to be blue and italic
text.setStyle(blueItalic);

// Overwrites blue and italic style to be red, bold, underlined, and strikethrough
text.withStyle(red).withStyle(bold).withStyle(doubleLines);
```

> This results in red, bold text with two lines as shown in the [image][red_hello]

Reviewer: Image missing `!` in front. Also doesn't need to be a sentence.

Chaining Components
-------------------

`MutableComponent` can have additional components chained as siblings with `MutableComponent#append`. Chained components can be retrieved with `MutableComponent#getSiblings`.

> `Component` stores its siblings like a tree, and its style is merged with those of siblings as shown in the [image][tree].

Reviewer: ```suggestion
`Component` stores its siblings like a tree and traversed in preorder; the parent style is merged with those of its siblings.
```

Image can be below it.

The tree is traversed in preorder.

> Below example will create a component with the same structure as the above image:

Reviewer: ```suggestion
The code below will create a component with the same structure in the above example:
```

```java
// Create text only components
MutableComponent first = Component.literal("first ");
MutableComponent second = Component.literal("second ");
MutableComponent third = Component.literal("third ");
MutableComponent fourth = Component.literal("fourth ");
MutableComponent fifth = Component.literal("fifth ");
MutableComponent sixth = Component.literal("sixth ");
MutableComponent seventh = Component.literal("seventh ");

// Create components with style
MutableComponent red = Component.litearl("red ").withStyle(Style.EMPTY.withColor(0xFF0000));
MutableComponent blue = Component.literal("blue ").withStyle(Style.EMPTY.withColor(0x0000FF));
MutableComponent bold = Component.literal("bold ").withStyle(Style.EMPTY.withBold(true));

// Structure created components in the same way as the image
red.append(first).append(blue).append(seventh);
blue.append(second).append(third).append(bold);
bold.append(fourth).append(fifth).append(sixth);
```

> [Here's how it looks like in-game][style_annotated]

Reviewer: Missing `!` again. Text should be alt text for image.

`ComponentContents`
-------------------

Component's text contents are handled by `ComponentContents`, it holds data and defines how to represent it as text.

> Notably, subtype `TranslatableContents` not only supports localization but also text formatting.

Reviewer: This is again unnecessary since you're just separating text formatting from the contents itself. Either more information is on the contents, or just include the first sentence as part of the general description of components.

### Text Formatting

Text formatting is the process of inserting data as text into predefined larger text.
It can be used for displaying coordinate with x, y, z annotation, showing number with respectful units alongside, etc.
Usually special notation called **format specifiers** are used for indicating where a text can be inserted into.

> `TranslatableContents` uses two types of format specifiers: `%s` and `%1$s`, `%2$s`, `%3$s`.

Reviewer: ```suggestion
`TranslatableContents` uses two types of format specifiers: `%s` and `%n$s`.
```

Its `args` are inserted in place of format specifiers.

> This feature is useful as order of information in various languages varies.

Reviewer: Unnecessary sentences.

> `%s` is replaced with elements of `args` in order they appear, i.e., First `%s` is replaced with a first element of `args`, and so on.

Reviewer: What is `args`? I don't remember this ever being defined.
